### PR TITLE
[GatherElements]: Implement verification, shape inference,  code gen

### PR DESCRIPTION
**Pull request by etiotto**
Implement support for the ONNX `GatherElements` operator:

- [x] verification code (diagnose operator constraints)
- [x] shape inference with helper
- [x] codegen support
- [x] add lit tests to verify constraint diagnostics 
- [x] add lit test to verify code generation
- [x] enable end-to-end test (backend test)

Signed-off-by: Ettore Tiotto <etiotto@ca.ibm.com>

### CODE REVIEW

File: src/Conversion/ONNXToKrnl/Tensor/Gather.cpp [hunk 1]
```diff
@@ -41,18 +41,28 @@ struct ONNXGatherOpLowering : public ConversionPattern {
     Value alloc = insertAllocAndDeallocSimple(
         rewriter, op, outputMemRefType, loc, shapeHelper.dimsForOutput());
 
-    // Save axis and rank info.
+    // Operands and attributes.
+    Value data = operandAdaptor.data();
+    Value indices = operandAdaptor.indices();
     int64_t axisLit = gatherOp.axis();
-    int64_t dataRank = shapeHelper.dataDims.size();
```
Comment (etiotto) [inline src/Conversion/ONNXToKrnl/Tensor/Gather.cpp:46]: I have simplified the shapehelper implementation, to make it consistent with other shape helpers.

File: src/Conversion/ONNXToKrnl/Tensor/ScatterElements.cpp [hunk 2]
```diff
@@ -19,17 +19,6 @@ using namespace mlir;
 
 namespace onnx_mlir {
 
-// Returns true if all the indices are known to be positive and false otherwise.
-static bool indicesArePositiveConstants(Value indices) {
```
Comment (etiotto) [inline src/Conversion/ONNXToKrnl/Tensor/ScatterElements.cpp:23]: Placed this utility function in ONNXtoKernlCommon.cpp so that it can be used by Gather, GatherElements, and ScatterElements during code generation (conversion from ONNX to Krnl).

File: src/Dialect/ONNX/ShapeInference/Gather.cpp [hunk 3]
```diff
@@ -14,59 +18,21 @@ using namespace mlir;
...
   MemRefBoundsIndexCapture dataBounds(operandAdaptor.data());
   MemRefBoundsIndexCapture indicesBounds(operandAdaptor.indices());
+  DimsExpr dataDims, indicesDims;
   dataBounds.getDimList(dataDims);
   indicesBounds.getDimList(indicesDims);
 
-  // Read constant 'axis' attribute and normalize when negative.
+  int64_t dataRank = dataDims.size();
   int64_t axisIndex = op->axis();
-  // The 'axis' value must be in [-rank, rank-1].
```
Comment (etiotto) [inline src/Dialect/ONNX/ShapeInference/Gather.cpp:42]: Simplified to make more consistent with other shape helpers.  The auxiliary vector (dataDims) now computed in the ONNXToKrnl conversion code.

File: src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp [hunk 4]
```diff
@@ -296,21 +298,6 @@ struct ONNXMatMulOpShapeHelper : public ONNXOpShapeHelper<mlir::ONNXMatMulOp> {
       bPadDims; // When true, that dim was padded.
 };
 
-// Shape for Gather.
-struct ONNXGatherOpShapeHelper : public ONNXOpShapeHelper<mlir::ONNXGatherOp> {
```
Comment (etiotto) [inline src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp:300]: Now generated via `DECLARE_SHAPE_HELPER`

File: test/mlir/onnx/onnx_lowering.mlir [hunk 5]
```diff
@@ -2484,31 +2484,31 @@ func @test_resize2(%arg0 : tensor<3x4xf32>) -> tensor<*xf32> {
...
 
-
 //-----
+
   func @test_gather_scalar(%arg0: tensor<4xi64>, %arg1: tensor<i64>) -> tensor<i64> {
     %0 = "onnx.Gather"(%arg0, %arg1) {axis = 0 : si64} : (tensor<4xi64>, tensor<i64>) -> tensor<i64>
     return %0 : tensor<i64>
 // CHECK-LABEL:  @test_gather_scalar
 // CHECK-SAME:   ([[PARAM_0_:%.+]]: memref<4xi64>, [[PARAM_1_:%.+]]: memref<i64>) -> memref<i64> {
-// CHECK-DAG:       [[CST_4_:%.+]] = arith.constant 4 : index
```
Comment (etiotto) [inline test/mlir/onnx/onnx_lowering.mlir:2494]: Codegen essentially the same, some constant moved around.

File: test/mlir/onnx/onnx_shape_inference.mlir [hunk 6]
```diff
@@ -1564,7 +1564,7 @@ func @test_gather_negative_axis(%arg0 : tensor<3x3xf32>, %arg1 : tensor<1x2xi64>
   "std.return"(%0) : (tensor<*xf32>) -> ()
 
   // CHECK-LABEL: test_gather_negative_axis
-  // CHECK: [[RES:%.+]] = "onnx.Gather"(%arg0, %arg1) {axis = 1 : si64} : (tensor<3x3xf32>, tensor<1x2xi64>) -> tensor<3x1x2xf32>
+  // CHECK: [[RES:%.+]] = "onnx.Gather"(%arg0, %arg1) {axis = -1 : si64} : (tensor<3x3xf32>, tensor<1x2xi64>) -> tensor<3x1x2xf32>
```
Comment (etiotto) [inline test/mlir/onnx/onnx_shape_inference.mlir:1567]: shapeHelper used to change the negative axis value. Now we are leaving the axis attribute as it was on the original graph, and dealing with the semantics associated with a negative axis value during codegen (same approach as in ScatterElements, GatherElements, and ScatterND).

**Review by etiotto — COMMENTED** [review]
Adding some notes to help code reviewers understand changes more easily.

File: src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp [hunk 7]
```diff
@@ -111,6 +111,18 @@ bool hasAllScalarValues(ArrayRef<Value> values) {
...
 
+/// Check if the value is a KrnlGlobalOp with a dense attribute of positive
+/// constants.
+bool indicesArePositiveConstants(Value indices) {
+  DenseElementsAttr valueAttribute =
+      krnl::getDenseElementAttributeFromKrnlValue(indices);
+  if (!valueAttribute)
+    return false;
+
+  return llvm::all_of(valueAttribute.getValues<IntegerAttr>(),
```
Comment (tungld) [inline src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp:122]: It seems you assume that the elements of DenseElementsAttr are IntegerAttr. Prefer to mention that in the function's comment or explicitly handle it.

File: src/Conversion/ONNXToKrnl/Tensor/Gather.cpp [hunk 8]
```diff
@@ -41,18 +41,28 @@ struct ONNXGatherOpLowering : public ConversionPattern {
...
-    int64_t indicesRank = shapeHelper.indicesDims.size();
-    int64_t outputRank = shapeHelper.dimsForOutput().size();
+    int64_t dataRank = data.getType().cast<ShapedType>().getRank();
```
Comment (tungld) [inline src/Conversion/ONNXToKrnl/Tensor/Gather.cpp:48]: `ShapedType` can be unranked, e.g. `TensorType<*xtype>`, and `getRank()` would fail. Maybe, `MemRefType` is better.

File: src/Conversion/ONNXToKrnl/Tensor/GatherElements.cpp [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    // Insert an allocation and deallocation for the result of this operation.
+    MemRefType outputMemRefType = convertToMemRefType(*op->result_type_begin());
+    Value output = insertAllocAndDeallocSimple(
+        rewriter, op, outputMemRefType, loc, shapeHelper.dimsForOutput());
+
+    // Operands and attributes.
+    Value data = operandAdaptor.data();
+    Value indices = operandAdaptor.indices();
+    int64_t axis = gatherElementsOp.axis();
+    int64_t dataRank = data.getType().cast<ShapedType>().getRank();
```
Comment (tungld) [inline src/Conversion/ONNXToKrnl/Tensor/GatherElements.cpp:48]: `ShapedType` can be unranked, e.g. `TensorType<*xtype>`, and `getRank()` would fail. Maybe, `MemRefType` is better.

File: src/Conversion/ONNXToKrnl/Tensor/GatherElements.cpp [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+          // Compute index = indices[i][j]...[n]
+          Value indexVal = createKrnl.loadIE(indices, accessFct);
+          IndexExpr index = NonAffineIndexExpr(indexVal);
+
+          // When index may be negative, add axis dim to it.
+          if (!indicesArePositives) {
+            LiteralIndexExpr zero(0);
+            SymbolIndexExpr axisDim(dataDims[axis]);
+            index = index.selectOrSelf(index < zero, index + axisDim);
```
Comment (tungld) [inline src/Conversion/ONNXToKrnl/Tensor/GatherElements.cpp:89]: I agree that this generally works for all cases. But, it seems that `indices` is often a constant tensor. Should we do constant propagation in advance to produce a new constant tensor whose values are not negative, so the loop here would be efficient. It's fine if you want to do this in another PR for both Gather and GatherElements.

File: src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp [hunk 11]
```diff
@@ -72,12 +72,13 @@ struct OnnxToKrnlBuilder : public OnnxBuilder {
...
 
-/// Check is all dimensions are known at compile time.
-bool hasAllConstantDimensions(MemRefType type);
-
-/// Check is all operands are scalar values at compile time.
+/// Check if all operands are scalar values at compile time.
 bool hasAllScalarValues(ArrayRef<Value> values);
 
+/// Check if the value is a KrnlGlobalOp with a dense attribute of positive
+/// constants.
```
Comment (tungld) [inline src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp:79]: It seems value 0 is OK, so `positive constants` => `non-negative constants`.

Comment (etiotto) [reply src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp:79]: Renamed function to `indicesAreNonNegative` and adjusted comment.

File: src/Conversion/ONNXToKrnl/Tensor/Gather.cpp (hunk 8 above)
Comment (etiotto) [reply src/Conversion/ONNXToKrnl/Tensor/Gather.cpp:48]: Good point!  Using MemrefType.

File: src/Conversion/ONNXToKrnl/Tensor/GatherElements.cpp (hunk 9 above)
Comment (etiotto) [reply src/Conversion/ONNXToKrnl/Tensor/GatherElements.cpp:48]: Using MemrefType.

File: src/Conversion/ONNXToKrnl/Tensor/GatherElements.cpp (hunk 10 above)
Comment (etiotto) [reply src/Conversion/ONNXToKrnl/Tensor/GatherElements.cpp:89]: OK. The code here is conservative. We can consider that in a different PR.

File: src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp (hunk 7 above)
Comment (etiotto) [reply src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp:122]: Done.

**Review by tungld — APPROVED** [review]
LGTM.
